### PR TITLE
persistence: fix premature reset of the 'writeInProgress' flag in cas…

### PR DESCRIPTION
**Pull request by RibleStrype**
A fix for
>Possible race condition in Akka Persistence that can lead to lost events when using persistAsync() #28629

The idea is not to reset the `writeInProgress` flag prematurely before the actor gets stopped by `WriteMessageFailure`.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by johanandren** [comment]
OK TO TEST

**Review by johanandren — COMMENTED** [review]
It would be good with some form of test repeating the issue, I understand it's a race but perhaps it would be possible using `akka.persistence.journal.SteppingInmemJournal` and controlling the journal carefully?


The `WriteMessagesFailed` is internal so can be filtered out of the bin comp. checks (see CONTRIBUTING.md for details)

**Comment by RibleStrype** [comment]
@johanandren thanks for the feedback, will try to come up with smth

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by johanandren** [comment]
I have manually verified that the CLA is signed. 

The reason why the CLA checker is complaining could be that you are using a different email in the commits than the one you used to sign the CLA.

**Review by johanandren — APPROVED** [review]
LGTM

I guess a repeater test turned out to be hard to achieve?

**Comment by RibleStrype** [comment]
yes, I haven’t yet found a way to reproduce the issue deterministically

**Review by patriknw — APPROVED** [review]
LGTM, makes sense

**Comment by patriknw** [comment]
@RibleStrype We need the CLA validator to pass before we can merge this. Could you squash amend the commits and use the email/user that you used for signing the CLA?

**Comment by RibleStrype** [comment]
Sorry for the late response, I’m on vacation right now without an access to computer. I can do it on Monday, hope that’s OK

**Comment by patriknw** [comment]
ok, thanks

**Comment by RibleStrype** [comment]
@patriknw I squashed the commits. The CLA is signed now

**Comment by RibleStrype** [comment]
could someone please help me understand why the Travis CI build is failing? I can't reproduce these binary incompatibility errors locally

**Comment by johanandren** [comment]
It's the binary compatibility check that is failing (mentioned here: https://github.com/akka/akka/blob/master/CONTRIBUTING.md#binary-compatibility )

```
[error] akka-persistence: Failed binary compatibility check against com.typesafe.akka:akka-persistence_2.12:2.5.27! Found 1 potential problems (filtered 319)
309[error]  * method unapply(akka.persistence.JournalProtocol#WriteMessagesFailed)scala.Option in object akka.persistence.JournalProtocol#WriteMessagesFailed has a different generic signature in current version, where it is (Lakka/persistence/JournalProtocol$WriteMessagesFailed;)Lscala/Option<Lscala/Tuple2<Ljava/lang/Throwable;Ljava/lang/Object;>;>; rather than (Lakka/persistence/JournalProtocol$WriteMessagesFailed;)Lscala/Option<Ljava/lang/Throwable;>;. See https://github.com/lightbend/mima#incompatiblesignatureproblem
310[error]    filter with: ProblemFilters.exclude[IncompatibleSignatureProblem]("akka.persistence.JournalProtocol#WriteMessagesFailed.unapply")
```

Since this is an internal message and not anything user or journal implementations interact with it should be safe to filter that in a `src/main/mima-filters/2.6.4.backwards.excludes/28629-writeInProgress-race.excludes`

**Comment by johanandren** [comment]
Mima passing both for 2.12 and 2.13 locally for me now at least, I hope travis will agree.

**Comment by johanandren** [comment]
The reason it doesn't show locally, master has been upgraded to mima 0.7.0, and the PR validation does a fetch merge while, at least my own, running it locally uses the 0.6.4 version from when the branch was created. If you rebase on top of current master you will see the same issue.

Still doesn't explain why it reports something that is already filtered though.

**Comment by johanandren** [comment]
Ah, it actually detects an additional incompatibiliey with mima 0.7.0, adding:
`ProblemFilters.exclude[IncompatibleSignatureProblem]("akka.persistence.JournalProtocol#WriteMessagesFailed.unapply")` should do the trick

**Comment by RibleStrype** [comment]
@johanandren yes, I have noticed it and pushed the fix, let's see how it goes

**Comment by RibleStrype** [comment]
@johanandren, @patriknw thanks a lot for your help

**Comment by johanandren** [comment]
Thanks a lot for you contribution! :)

**Pull request merged**
